File: api/app/routers/chats.py
"""Endpoints for chats, members and chat management operations."""

from __future__ import annotations

import json
from typing import Any

from fastapi import APIRouter, HTTPException

from ..db import execute, fetch_all, fetch_one
from ..models import (
    SetChatAliasIn,
    SetChatTitleIn,
    SetChatDescriptionIn,
    CreateChatInviteLinkIn,
    EditChatInviteLinkIn,
    RevokeChatInviteLinkIn,
    CreateSubscriptionInviteLinkIn,
    EditChatSubscriptionInviteLinkIn,
)
from ..telegram_client import (
    TelegramError,
    ban_chat_member,
    get_chat,
    get_chat_member,
    get_chat_member_count,
    pin_chat_message,
    promote_chat_member,
    restrict_chat_member,
    unban_chat_member,
    unpin_chat_message,
    set_chat_title,
    set_chat_description,
    delete_chat_photo,
    leave_chat,
    unpin_all_chat_messages,
    create_chat_invite_link,
    edit_chat_invite_link,
    revoke_chat_invite_link,
    export_chat_invite_link,
    create_chat_subscription_invite_link,
    edit_chat_subscription_invite_link,
)
from ..utils import resolve_bot_context

router = APIRouter(prefix="/v1/chats", tags=["chats"])


@router.get("")
async def list_chats_api(
    bot_id: int | None = None,
    chat_type: str | None = None,
    limit: int = 100,
    offset: int = 0,
) -> dict[str, Any]:
    where: list[str] = []
    values: list[Any] = []
    if bot_id is not None:
        where.append("bot_id = %s")
        values.append(bot_id)
    if chat_type:
        where.append("type = %s")
        values.append(chat_type)

    where_sql = f"WHERE {' AND '.join(where)}" if where else ""
    rows = await fetch_all(
        f"""
        SELECT *
        FROM chats
        {where_sql}
        ORDER BY updated_at DESC, created_at DESC
        LIMIT %s OFFSET %s
        """,
        [*values, limit, offset],
    )
    return {"items": rows, "count": len(rows)}


@router.put("/{chat_id}/alias")
async def set_chat_alias_api(chat_id: str, payload: SetChatAliasIn) -> dict[str, Any]:
    row = await execute(
        """
        UPDATE chats
        SET alias = %s,
            updated_at = NOW()
        WHERE chat_id = %s
        """,
        [payload.alias.strip(), chat_id],
    )
    if row is None:
        # execute() doesn't return rowcount; verify existence explicitly.
        exists = await fetch_one("SELECT chat_id FROM chats WHERE chat_id = %s", [chat_id])
        if not exists:
            raise HTTPException(status_code=404, detail="chat not found")

    chat = await fetch_one("SELECT * FROM chats WHERE chat_id = %s", [chat_id])
    return {"chat": chat}


@router.get("/by-alias/{alias}")
async def get_chat_by_alias_api(alias: str) -> dict[str, Any]:
    row = await fetch_one("SELECT * FROM chats WHERE alias = %s", [alias])
    if not row:
        raise HTTPException(status_code=404, detail="chat alias not found")
    return {"chat": row}


@router.get("/{chat_id}/history")
async def get_chat_history_api(
    chat_id: str,
    bot_id: int | None = None,
    limit: int = 100,
    offset: int = 0,
) -> dict[str, Any]:
    where = ["chat_id = %s"]
    values: list[Any] = [chat_id]
    if bot_id is not None:
        where.append("bot_id = %s")
        values.append(bot_id)

    rows = await fetch_all(
        f"""
        SELECT *
        FROM messages
        WHERE {' AND '.join(where)}
        ORDER BY created_at DESC
        LIMIT %s OFFSET %s
        """,
        [*values, limit, offset],
    )
    return {"items": rows, "count": len(rows)}


@router.get("/{chat_id}/members")
async def list_chat_members_api(
    chat_id: str,
    bot_id: int | None = None,
    limit: int = 200,
    offset: int = 0,
) -> dict[str, Any]:
    where = ["cm.chat_id = %s"]
    values: list[Any] = [chat_id]
    if bot_id is not None:
        where.append("cm.bot_id = %s")
        values.append(bot_id)

    rows = await fetch_all(
        f"""
        SELECT
            cm.*,
            u.username,
            u.first_name,
            u.last_name,
            u.is_premium
        FROM chat_members cm
        LEFT JOIN users u ON u.user_id = cm.user_id
        WHERE {' AND '.join(where)}
        ORDER BY cm.updated_at DESC
        LIMIT %s OFFSET %s
        """,
        [*values, limit, offset],
    )
    return {"items": rows, "count": len(rows)}


@router.get("/{chat_id}")
async def get_chat_api(chat_id: str, bot_id: int | None = None) -> dict[str, Any]:
    """Get chat details from Telegram API."""
    try:
        bot_token, resolved_bot_id = await resolve_bot_context(bot_id)
        result = await get_chat({"chat_id": chat_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    await execute(
        """
        INSERT INTO chats (chat_id, type, title, username, description, bot_id)
        VALUES (%s, %s, %s, %s, %s, %s)
        ON CONFLICT (chat_id) DO UPDATE
        SET type = EXCLUDED.type,
            title = EXCLUDED.title,
            username = EXCLUDED.username,
            description = EXCLUDED.description,
            bot_id = COALESCE(EXCLUDED.bot_id, chats.bot_id),
            updated_at = NOW()
        """,
        [
            str(result.get("id") or chat_id),
            result.get("type"),
            result.get("title"),
            result.get("username"),
            result.get("description"),
            resolved_bot_id,
        ],
    )

    return {"chat": result}


@router.get("/{chat_id}/members/{user_id}")
async def get_chat_member_api(chat_id: str, user_id: int, bot_id: int | None = None) -> dict[str, Any]:
    """Get member details from Telegram API."""
    try:
        bot_token, resolved_bot_id = await resolve_bot_context(bot_id)
        result = await get_chat_member({"chat_id": chat_id, "user_id": user_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    await execute(
        """
        INSERT INTO chat_members (chat_id, user_id, bot_id, status, last_seen_at, metadata)
        VALUES (%s, %s, %s, %s, NOW(), %s::jsonb)
        ON CONFLICT (chat_id, user_id) DO UPDATE
        SET bot_id = COALESCE(EXCLUDED.bot_id, chat_members.bot_id),
            status = EXCLUDED.status,
            last_seen_at = NOW(),
            metadata = EXCLUDED.metadata,
            updated_at = NOW()
        """,
        [chat_id, str(user_id), resolved_bot_id, result.get("status"), json.dumps(result)],
    )

    return {"member": result}


@router.get("/{chat_id}/members/count")
async def get_chat_member_count_api(chat_id: str, bot_id: int | None = None) -> dict[str, Any]:
    """Get number of chat members from Telegram API."""
    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await get_chat_member_count({"chat_id": chat_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"count": result}


@router.post("/{chat_id}/pin/{message_id}")
async def pin_message_api(chat_id: str, message_id: int, bot_id: int | None = None) -> dict[str, Any]:
    """Pin a message in chat."""
    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await pin_chat_message(
            {
                "chat_id": chat_id,
                "message_id": message_id,
            },
            bot_token=bot_token,
        )
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.delete("/{chat_id}/pin/{message_id}")
async def unpin_message_api(chat_id: str, message_id: int, bot_id: int | None = None) -> dict[str, Any]:
    """Unpin a message in chat."""
    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await unpin_chat_message(
            {
                "chat_id": chat_id,
                "message_id": message_id,
            },
            bot_token=bot_token,
        )
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.get("/{chat_id}/stored")
async def get_stored_chat_api(chat_id: str) -> dict[str, Any]:
    """Get chat data from local DB (captured from webhooks)."""
    row = await fetch_one("SELECT * FROM chats WHERE chat_id = %s", [chat_id])
    if not row:
        raise HTTPException(status_code=404, detail="chat not found in local db")
    return {"chat": row}


@router.get("/{chat_id}/stored-users")
async def list_chat_users_api(chat_id: str, limit: int = 100, offset: int = 0) -> dict[str, Any]:
    """List users from DB that were seen in this chat."""
    rows = await fetch_all(
        """
        SELECT DISTINCT u.*
        FROM users u
        JOIN webhook_updates wu ON wu.chat_id = %s AND wu.user_id = u.user_id
        ORDER BY u.updated_at DESC
        LIMIT %s OFFSET %s
        """,
        [chat_id, limit, offset],
    )
    return {"items": rows, "count": len(rows)}


# === Chat management ===


@router.post("/{chat_id}/members/{user_id}/ban")
async def ban_member_api(
    chat_id: str,
    user_id: int,
    until_date: int | None = None,
    revoke_messages: bool = False,
    bot_id: int | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "chat_id": chat_id,
        "user_id": user_id,
        "revoke_messages": revoke_messages,
    }
    if until_date:
        payload["until_date"] = until_date

    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await ban_chat_member(payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    return {"ok": True, "result": result}


@router.post("/{chat_id}/members/{user_id}/unban")
async def unban_member_api(
    chat_id: str,
    user_id: int,
    only_if_banned: bool = True,
    bot_id: int | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "chat_id": chat_id,
        "user_id": user_id,
        "only_if_banned": only_if_banned,
    }

    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await unban_chat_member(payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    return {"ok": True, "result": result}


@router.post("/{chat_id}/members/{user_id}/restrict")
async def restrict_member_api(
    chat_id: str,
    user_id: int,
    permissions: dict[str, bool],
    until_date: int | None = None,
    bot_id: int | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "chat_id": chat_id,
        "user_id": user_id,
        "permissions": permissions,
    }
    if until_date:
        payload["until_date"] = until_date

    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await restrict_chat_member(payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    return {"ok": True, "result": result}


@router.post("/{chat_id}/members/{user_id}/promote")
async def promote_member_api(
    chat_id: str,
    user_id: int,
    is_anonymous: bool = False,
    can_manage_chat: bool = False,
    can_post_messages: bool = False,
    can_edit_messages: bool = False,
    can_delete_messages: bool = False,
    can_manage_video_chats: bool = False,
    can_restrict_members: bool = False,
    can_promote_members: bool = False,
    can_change_info: bool = False,
    can_invite_users: bool = False,
    can_pin_messages: bool = False,
    bot_id: int | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "chat_id": chat_id,
        "user_id": user_id,
        "is_anonymous": is_anonymous,
        "can_manage_chat": can_manage_chat,
        "can_post_messages": can_post_messages,
        "can_edit_messages": can_edit_messages,
        "can_delete_messages": can_delete_messages,
        "can_manage_video_chats": can_manage_video_chats,
        "can_restrict_members": can_restrict_members,
        "can_promote_members": can_promote_members,
        "can_change_info": can_change_info,
        "can_invite_users": can_invite_users,
        "can_pin_messages": can_pin_messages,
    }

    try:
        bot_token, _ = await resolve_bot_context(bot_id)
        result = await promote_chat_member(payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    return {"ok": True, "result": result}


# === Batch 7: Chat Administration ===


@router.put("/{chat_id}/title")
async def set_chat_title_api(chat_id: str, payload: SetChatTitleIn) -> dict[str, Any]:
    """Установить название чата."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    try:
        result = await set_chat_title({"chat_id": chat_id, "title": payload.title}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.put("/{chat_id}/description")
async def set_chat_description_api(chat_id: str, payload: SetChatDescriptionIn) -> dict[str, Any]:
    """Установить описание чата."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    telegram_payload: dict[str, Any] = {"chat_id": chat_id}
    if payload.description is not None:
        telegram_payload["description"] = payload.description
    try:
        result = await set_chat_description(telegram_payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.delete("/{chat_id}/photo")
async def delete_chat_photo_api(chat_id: str, bot_id: int | None = None) -> dict[str, Any]:
    """Удалить фото чата."""
    bot_token, _ = await resolve_bot_context(bot_id)
    try:
        result = await delete_chat_photo({"chat_id": chat_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.post("/{chat_id}/leave")
async def leave_chat_api(chat_id: str, bot_id: int | None = None) -> dict[str, Any]:
    """Выйти из чата."""
    bot_token, _ = await resolve_bot_context(bot_id)
    try:
        result = await leave_chat({"chat_id": chat_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.post("/{chat_id}/unpin-all")
async def unpin_all_chat_messages_api(chat_id: str, bot_id: int | None = None) -> dict[str, Any]:
    """Открепить все сообщения в чате."""
    bot_token, _ = await resolve_bot_context(bot_id)
    try:
        result = await unpin_all_chat_messages({"chat_id": chat_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.post("/{chat_id}/invite-links")
async def create_chat_invite_link_api(chat_id: str, payload: CreateChatInviteLinkIn) -> dict[str, Any]:
    """Создать пригласительную ссылку."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    telegram_payload: dict[str, Any] = {"chat_id": chat_id}
    if payload.name:
        telegram_payload["name"] = payload.name
    if payload.expire_date is not None:
        telegram_payload["expire_date"] = payload.expire_date
    if payload.member_limit is not None:
        telegram_payload["member_limit"] = payload.member_limit
    if payload.creates_join_request:
        telegram_payload["creates_join_request"] = payload.creates_join_request
    try:
        result = await create_chat_invite_link(telegram_payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.put("/{chat_id}/invite-links")
async def edit_chat_invite_link_api(chat_id: str, payload: EditChatInviteLinkIn) -> dict[str, Any]:
    """Редактировать пригласительную ссылку."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    telegram_payload: dict[str, Any] = {
        "chat_id": chat_id,
        "invite_link": payload.invite_link,
    }
    if payload.name is not None:
        telegram_payload["name"] = payload.name
    if payload.expire_date is not None:
        telegram_payload["expire_date"] = payload.expire_date
    if payload.member_limit is not None:
        telegram_payload["member_limit"] = payload.member_limit
    if payload.creates_join_request is not None:
        telegram_payload["creates_join_request"] = payload.creates_join_request
    try:
        result = await edit_chat_invite_link(telegram_payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.delete("/{chat_id}/invite-links")
async def revoke_chat_invite_link_api(chat_id: str, payload: RevokeChatInviteLinkIn) -> dict[str, Any]:
    """Отозвать пригласительную ссылку."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    try:
        result = await revoke_chat_invite_link(
            {"chat_id": chat_id, "invite_link": payload.invite_link},
            bot_token=bot_token,
        )
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.post("/{chat_id}/export-invite-link")
async def export_chat_invite_link_api(chat_id: str, bot_id: int | None = None) -> dict[str, Any]:
    """Экспорт основной пригласительной ссылки."""
    bot_token, _ = await resolve_bot_context(bot_id)
    try:
        result = await export_chat_invite_link({"chat_id": chat_id}, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.post("/{chat_id}/subscription-links")
async def create_subscription_invite_link_api(chat_id: str, payload: CreateSubscriptionInviteLinkIn) -> dict[str, Any]:
    """Создать подписочную пригласительную ссылку (Bot API 7.9)."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    telegram_payload: dict[str, Any] = {
        "chat_id": chat_id,
        "subscription_period": payload.subscription_period,
        "subscription_price": payload.subscription_price,
    }
    if payload.name:
        telegram_payload["name"] = payload.name
    try:
        result = await create_chat_subscription_invite_link(telegram_payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}


@router.put("/{chat_id}/subscription-links")
async def edit_subscription_invite_link_api(chat_id: str, payload: EditChatSubscriptionInviteLinkIn) -> dict[str, Any]:
    """Редактировать подписочную пригласительную ссылку (Bot API 7.9)."""
    bot_token, _ = await resolve_bot_context(payload.bot_id)
    telegram_payload: dict[str, Any] = {
        "chat_id": chat_id,
        "invite_link": payload.invite_link,
    }
    if payload.name is not None:
        telegram_payload["name"] = payload.name
    try:
        result = await edit_chat_subscription_invite_link(telegram_payload, bot_token=bot_token)
    except TelegramError as exc:
        raise HTTPException(status_code=502, detail=str(exc))
    return {"ok": True, "result": result}

File: mcp/src/tools/chats.ts
import { z } from "zod";
import type { ToolDef, ApiRequestFn } from "../types.js";

/** Инструменты для работы с чатами, callback queries и администрирования */
export function register(apiRequest: ApiRequestFn): ToolDef[] {
  return [
    {
      name: "chats.get",
      description: "Получить информацию о чате от Telegram API (название, тип, описание).",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => {
        const qs = new URLSearchParams();
        if (params.bot_id !== undefined) qs.set("bot_id", String(params.bot_id));
        return apiRequest(`/v1/chats/${params.chat_id}${qs.toString() ? `?${qs.toString()}` : ""}`);
      },
    },
    {
      name: "chats.member",
      description: "Получить статус участника чата (admin, member, restricted и т.д.).",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        user_id: z.number().int(),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => {
        const qs = new URLSearchParams();
        if (params.bot_id !== undefined) qs.set("bot_id", String(params.bot_id));
        return apiRequest(`/v1/chats/${params.chat_id}/members/${params.user_id}${qs.toString() ? `?${qs.toString()}` : ""}`);
      },
    },
    {
      name: "chats.list",
      description: "Список чатов из локальной БД с фильтрацией по bot_id и типу.",
      parameters: z.object({
        bot_id: z.number().int().optional(),
        chat_type: z.string().optional(),
        limit: z.number().int().min(1).max(500).optional().default(100),
        offset: z.number().int().min(0).optional().default(0),
      }),
      execute: async (params) => {
        const qs = new URLSearchParams();
        if (params.bot_id !== undefined) qs.set("bot_id", String(params.bot_id));
        if (params.chat_type) qs.set("chat_type", params.chat_type);
        qs.set("limit", String(params.limit));
        qs.set("offset", String(params.offset));
        return apiRequest(`/v1/chats?${qs.toString()}`);
      },
    },
    {
      name: "chats.alias",
      description: "Установить алиас для чата в локальной БД.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        alias: z.string().min(2).max(120),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/alias`, {
        method: "PUT",
        body: JSON.stringify({ alias: params.alias }),
      }),
    },
    {
      name: "chats.history",
      description: "Получить историю сообщений чата из локальной БД.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
        limit: z.number().int().min(1).max(500).optional().default(100),
        offset: z.number().int().min(0).optional().default(0),
      }),
      execute: async (params) => {
        const qs = new URLSearchParams();
        if (params.bot_id !== undefined) qs.set("bot_id", String(params.bot_id));
        qs.set("limit", String(params.limit));
        qs.set("offset", String(params.offset));
        return apiRequest(`/v1/chats/${params.chat_id}/history?${qs.toString()}`);
      },
    },
    {
      name: "callbacks.answer",
      description: "Ответить на нажатие inline-кнопки (callback_query).",
      parameters: z.object({
        bot_id: z.number().int().optional(),
        callback_query_id: z.string(),
        text: z.string().optional().describe("Текст уведомления"),
        show_alert: z.boolean().optional().describe("Показать alert вместо toast"),
      }),
      execute: async (params) => apiRequest("/v1/callbacks/answer", {
        method: "POST",
        body: JSON.stringify(params),
      }),
    },
    {
      name: "callbacks.list",
      description: "Получить список полученных callback queries.",
      parameters: z.object({
        chat_id: z.string().optional(),
        user_id: z.string().optional(),
        answered: z.boolean().optional(),
        limit: z.number().int().min(1).max(500).optional().default(50),
        offset: z.number().int().min(0).optional().default(0),
      }),
      execute: async (params) => {
        const qs = new URLSearchParams();
        if (params.chat_id) qs.set("chat_id", params.chat_id);
        if (params.user_id) qs.set("user_id", params.user_id);
        if (params.answered !== undefined) qs.set("answered", String(params.answered));
        qs.set("limit", String(params.limit));
        qs.set("offset", String(params.offset));
        return apiRequest(`/v1/callbacks?${qs.toString()}`);
      },
    },

    // === Batch 7: Chat Administration ===
    {
      name: "chats.set_title",
      description: "Установить название чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
        title: z.string().max(128),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/title`, {
        method: "PUT",
        body: JSON.stringify({ bot_id: params.bot_id, title: params.title }),
      }),
    },
    {
      name: "chats.set_description",
      description: "Установить описание чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
        description: z.string().max(255).optional(),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/description`, {
        method: "PUT",
        body: JSON.stringify({ bot_id: params.bot_id, description: params.description }),
      }),
    },
    {
      name: "chats.delete_photo",
      description: "Удалить фото чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => {
        const qs = params.bot_id !== undefined ? `?bot_id=${params.bot_id}` : "";
        return apiRequest(`/v1/chats/${params.chat_id}/photo${qs}`, { method: "DELETE" });
      },
    },
    {
      name: "chats.leave",
      description: "Выйти из чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => {
        const qs = params.bot_id !== undefined ? `?bot_id=${params.bot_id}` : "";
        return apiRequest(`/v1/chats/${params.chat_id}/leave${qs}`, { method: "POST" });
      },
    },
    {
      name: "chats.unpin_all",
      description: "Открепить все сообщения в чате.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => {
        const qs = params.bot_id !== undefined ? `?bot_id=${params.bot_id}` : "";
        return apiRequest(`/v1/chats/${params.chat_id}/unpin-all${qs}`, { method: "POST" });
      },
    },
    {
      name: "chats.create_invite_link",
      description: "Создать пригласительную ссылку для чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
        name: z.string().max(32).optional(),
        expire_date: z.number().int().optional(),
        member_limit: z.number().int().min(1).max(99999).optional(),
        creates_join_request: z.boolean().optional(),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/invite-links`, {
        method: "POST",
        body: JSON.stringify(params),
      }),
    },
    {
      name: "chats.edit_invite_link",
      description: "Редактировать пригласительную ссылку чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        invite_link: z.string(),
        bot_id: z.number().int().optional(),
        name: z.string().max(32).optional(),
        expire_date: z.number().int().optional(),
        member_limit: z.number().int().min(1).max(99999).optional(),
        creates_join_request: z.boolean().optional(),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/invite-links`, {
        method: "PUT",
        body: JSON.stringify(params),
      }),
    },
    {
      name: "chats.revoke_invite_link",
      description: "Отозвать пригласительную ссылку чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        invite_link: z.string(),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/invite-links`, {
        method: "DELETE",
        body: JSON.stringify(params),
      }),
    },
    {
      name: "chats.export_invite_link",
      description: "Экспорт основной пригласительной ссылки чата.",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
      }),
      execute: async (params) => {
        const qs = params.bot_id !== undefined ? `?bot_id=${params.bot_id}` : "";
        return apiRequest(`/v1/chats/${params.chat_id}/export-invite-link${qs}`, { method: "POST" });
      },
    },
    {
      name: "chats.create_subscription_invite_link",
      description: "Создать подписочную пригласительную ссылку (Bot API 7.9).",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        bot_id: z.number().int().optional(),
        name: z.string().max(32).optional(),
        subscription_period: z.number().int().describe("Период подписки в секундах"),
        subscription_price: z.number().int().describe("Цена подписки в Stars"),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/subscription-links`, {
        method: "POST",
        body: JSON.stringify(params),
      }),
    },
    {
      name: "chats.edit_subscription_invite_link",
      description: "Редактировать подписочную пригласительную ссылку (Bot API 7.9).",
      parameters: z.object({
        chat_id: z.union([z.string(), z.number()]),
        invite_link: z.string(),
        bot_id: z.number().int().optional(),
        name: z.string().max(32).optional(),
      }),
      execute: async (params) => apiRequest(`/v1/chats/${params.chat_id}/subscription-links`, {
        method: "PUT",
        body: JSON.stringify(params),
      }),
    },
  ];
}
